### ISSUE TITLE
IDE0044 makes [ThreadStatic] statics readonly

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#42760]
**Version Used**: 
C# Tools 3.6.0-2.20167.1+926207070b0c62d6a76aec4627ec26dee17870bb

**Steps to Reproduce**:
```C#
using System;

class Program
{
    [ThreadStatic]
    private static object t_obj;

    static void Main() { }
}
``` 

**Expected Behavior**:
IDE0044 doesn't trigger.

**Actual Behavior**:
IDE0044 triggers and makes t_obj be readonly, which doesn't make sense for a [ThreadStatic].

**Comment by sharwell** [comment]
❔ If nothing is ever written to the field, does it make a difference?

I'm a bit unclear regarding the runtime semantics for this case.

**Comment by stephentoub** [comment]
It's just nonsensical.  The purpose of a [ThreadStatic] is to contain multiple values, one per thread, but a static cctor only possibly runs on one thread.  A [ThreadStatic] that's readonly then shouldn't be [ThreadStatic].  My point is the combination of [ThreadStatic] + readonly means something is wrong.

**Comment by sharwell** [comment]
Would this be another case where the code fix could be updated to simply ignore the case, even though IDE0044 is still reported? The user is alerted to the fact that the field is not being written to, but must manually decide the appropriate step(s) to correct it?

Alternatively, it might make more sense to just add a new analyzer that reports a warning if a field is both `[ThreadStatic]` and `readonly`. The code fix for IDE0044 keeps doing the same thing, and leaves the user with a more specific diagnostic regarding the two of them used together.

**Comment by stephentoub** [comment]
> Would this be another case where the code fix could be updated to simply ignore the case, even though IDE0044 is still reported?

That, or a dedicated analyzer / or diagnostic ID/message, would be fine.

When you say code fix is updated to ignore that case, you mean the code fix simply isn't offered, rather than it being a nop, right?

**Comment by sharwell** [comment]
> ... or a dedicated analyzer / or diagnostic ID/message, would be fine.

I lean toward this being the correct approach. From both the compiler and runtime perspective, the use of both `[ThreadStatic]` and `readonly` is acceptable, just likely not want the user intended. We can create a new analyzer covering this case, either as part of roslyn or as part of roslyn-analyzers.

> When you say code fix is updated to ignore that case, you mean the code fix simply isn't offered, rather than it being a nop, right?

This is correct.

**Comment by Youssef1313** [comment]
@sharwell What was the design review outcome? to leave IDE0044 unchanged and create an additional analyzer or something else?

**Comment by sharwell** [comment]
@Youssef1313 The conclusion was to update IDE0044 to ignore fields marked with `[ThreadStatic]`.

**Comment by Youssef1313** [comment]
@sharwell Thanks. I'll work on it. Sounds like a very small fix.
